File: vision/src/main/java/com/dreampany/vision/ui/fragment/LiveTextOcrFragment.java
package com.dreampany.vision.ui.fragment;

import android.Manifest;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.CheckBox;
import android.widget.TextView;

import com.dreampany.frame.misc.ActivityScope;
import com.dreampany.frame.ui.fragment.BaseMenuFragment;
import com.dreampany.frame.util.NotifyUtil;
import com.dreampany.frame.util.TextUtil;
import com.dreampany.vision.R;
import com.dreampany.vision.databinding.FragmentLiveTextOcrBinding;
import com.dreampany.vision.ml.CameraSource;
import com.dreampany.vision.ml.CameraSourcePreview;
import com.dreampany.vision.ml.GraphicOverlay;
import com.dreampany.vision.ml.ocr.TextRecognitionProcessor;
import com.google.android.gms.common.annotation.KeepName;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * Created by Hawladar Roman on 9/28/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
@KeepName
@ActivityScope
public class LiveTextOcrFragment extends BaseMenuFragment {

    FragmentLiveTextOcrBinding binding;

    CameraSource source = null;
    CameraSourcePreview preview;
    GraphicOverlay overlay;
    TextView textView;
    CheckBox check;
    StringBuilder texts = new StringBuilder();

    @Inject
    public LiveTextOcrFragment() {

    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_live_text_ocr;
    }

    @Override
    public int getMenuId() {
        return R.menu.menu_live_text_ocr;
    }

    @Override
    public void onMenuCreated(@NotNull Menu menu) {
        MenuItem checkItem = menu.findItem(R.id.item_auto_collection);
        check = (CheckBox) checkItem.getActionView();
        check.setOnCheckedChangeListener((buttonView, isChecked) -> {
            String text = isChecked ? "All text collection is enabled" : "All text collection is disabled";
            NotifyUtil.shortToast(getContext(), text);
        });
    }

    @Override
    protected void onStartUi(@Nullable Bundle state) {
        initView();

        Dexter.withActivity(getParent())
                .withPermissions(Manifest.permission.CAMERA,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {
                        if (report.areAllPermissionsGranted()) {
                            createCameraSource();
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {

                    }
                }).check();
    }

    @Override
    protected void onStopUi() {
        if (source != null) {
            source.release();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        startCameraSource();
    }

    @Override
    public void onPause() {
        preview.stop();
        super.onPause();
    }

    @Override
    public boolean hasBackPressed() {
        done();
        return false;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.item_clear) {
            clear();
            return true;
        }
        if (item.getItemId() == R.id.item_done) {
            done();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void initView() {

        setTitle(TextUtil.getString(getContext(), R.string.detected_words, 0));
        binding = (FragmentLiveTextOcrBinding) super.binding;
        preview = binding.preview;
        overlay = binding.overlay;
        textView = binding.text;
    }

    private void createCameraSource() {
        if (source == null) {
            source = new CameraSource(getParent(), overlay);
        }
        source.setMachineLearningFrameProcessor(new TextRecognitionProcessor(this::updateTitle));
    }

    private void startCameraSource() {
        if (source != null) {
            try {
                if (preview == null) {
                    Timber.d("resume: Preview is null");
                }
                if (overlay == null) {
                    Timber.d("resume: graphOverlay is null");
                }
                preview.start(source, overlay);
            } catch (IOException e) {
                Timber.e(e, "Unable to start camera source.");
                source.release();
                source = null;
            }
        }
    }

    private void updateTitle(String text) {
        if (!check.isChecked()) {
            texts.setLength(0);
        }
        texts.append(text);
        String result = texts.toString();
        textView.setText(text);
        int words = TextUtil.getWordsCount(result);
        setTitle(TextUtil.getString(getContext(), R.string.detected_words, words));
    }

    private void clear() {
        setTitle(TextUtil.getString(getContext(), R.string.detected_words, 0));
        textView.setText(null);
        texts.setLength(0);
    }

    private void done() {
        getCurrentTask(false).setComment(texts.toString());
        forResult();
    }
}

File: frame/src/main/kotlin/com/dreampany/vision/ui/fragment/LiveTextOcrFragment.kt
package com.dreampany.vision.ui.fragment

import android.Manifest
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.CheckBox
import android.widget.TextView
import com.afollestad.assent.Permission
import com.afollestad.assent.runWithPermissions
import com.dreampany.frame.R
import com.dreampany.frame.data.model.Task
import com.dreampany.frame.databinding.FragmentLiveTextOcrBinding
import com.dreampany.frame.misc.ActivityScope
import com.dreampany.frame.ui.fragment.BaseMenuFragment
import com.dreampany.frame.util.NotifyUtil
import com.dreampany.frame.util.TextUtil
import com.dreampany.vision.ml.CameraSource
import com.dreampany.vision.ml.CameraSourcePreview
import com.dreampany.vision.ml.GraphicOverlay
import com.dreampany.vision.ml.ocr.TextRecognitionProcessor
import com.google.android.gms.common.annotation.KeepName
import timber.log.Timber
import java.io.IOException
import javax.inject.Inject

/**
 * Created by Roman-372 on 7/18/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */

@KeepName
@ActivityScope
class LiveTextOcrFragment @Inject constructor() : BaseMenuFragment() {

    private lateinit var bind: FragmentLiveTextOcrBinding
    private var source: CameraSource? = null
    private lateinit var preview: CameraSourcePreview
    private lateinit var overlay: GraphicOverlay
    private lateinit var textView: TextView
    private lateinit var check: CheckBox
    private val texts = StringBuilder()

    override fun getLayoutId(): Int {
        return R.layout.fragment_live_text_ocr
    }

    override fun getMenuId(): Int {
        return R.menu.menu_live_text_ocr
    }

    override fun onMenuCreated(menu: Menu, inflater: MenuInflater) {
        val checkItem = menu.findItem(R.id.item_auto_collection)
        check = checkItem.actionView as CheckBox
        check.setOnCheckedChangeListener { buttonView, isChecked ->
            val text =
                if (isChecked) "All text collection is enabled" else "All text collection is disabled"
            NotifyUtil.shortToast(context, text)
        }
    }

    override fun onStartUi(state: Bundle?) {
        initView()

        runWithPermissions(Permission.CAMERA, Permission.WRITE_EXTERNAL_STORAGE) {
            createCameraSource()
        }
    }

    override fun onStopUi() {
        source?.release()
    }

    override fun onResume() {
        super.onResume()
        startCameraSource()
    }

    override fun onPause() {
        preview.stop()
        super.onPause()
    }

    override fun hasBackPressed(): Boolean {
        done()
        return false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.item_clear) {
            clear()
            return true
        }
        if (item.itemId == R.id.item_done) {
            done()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initView() {

        setTitle(TextUtil.getString(context, R.string.detected_words, 0))
        bind = super.binding as FragmentLiveTextOcrBinding
        preview = bind.preview
        overlay = bind.overlay
        textView = bind.text
    }

    private fun createCameraSource() {
        if (source == null) {
            source = CameraSource(getParent(), overlay)
        }
        source?.setMachineLearningFrameProcessor(TextRecognitionProcessor(TextRecognitionProcessor.Callback {
            this.updateTitle(
                it
            )
        }))
    }

    private fun startCameraSource() {
        if (source != null) {
            try {
                if (preview == null) {
                    Timber.d("resume: Preview is null")
                }
                if (overlay == null) {
                    Timber.d("resume: graphOverlay is null")
                }
                preview.start(source, overlay)
            } catch (e: IOException) {
                Timber.e(e, "Unable to start camera source.")
                source?.release()
                source = null
            }

        }
    }

    private fun updateTitle(text: String) {
        if (!check.isChecked) {
            texts.setLength(0)
        }
        texts.append(text)
        val result = texts.toString()
        textView.text = text
        val words = TextUtil.getWordsCount(result)
        setTitle(TextUtil.getString(context, R.string.detected_words, words))
    }

    private fun clear() {
        setTitle(TextUtil.getString(context, R.string.detected_words, 0))
        textView.text = null
        texts.setLength(0)
    }

    private fun done() {
        getCurrentTask<Task<*>>(false)!!.comment = texts.toString()
        forResult()
    }
}